### PR TITLE
Fix (accounting): order of processed fee events

### PR DESCRIPTION
**Pull request by OjusWiZard**
## Checklist

- [x] Add prior acquisition of the assets that are swapped in paraswap test
- [x] Fix order of processed events when fee asset == in asset
- [x] Test if the final PnLs and cost basis are correct, for `include_fees_in_cost_basis` = True/False

### CODE REVIEW

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 1]
```diff
@@ -114,8 +196,11 @@ def test_paraswap_swap_with_fee(accountant: 'Accountant'):
...
             pnl=PNL(taxable=ZERO, free=ZERO),
             cost_basis=None,
-            index=2,
             extra_data=extra_data,
         )]
     expected_processed_events[0].count_cost_basis_pnl = True  # since it's not settable at ctor
+    expected_processed_events[1].count_cost_basis_pnl = True  # since it's not settable at ctor
+    expected_processed_events[2].count_cost_basis_pnl = True  # since it's not settable at ctor
     assert pot.processed_events == expected_processed_events
+    assert pot.pnls[AccountingEventType.TRANSACTION_EVENT].taxable == FVal(swap_amount_str) + ONE
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:205]: I believe is clearer if you put where they come from

```suggestion
    assert pot.pnls[AccountingEventType.TRANSACTION_EVENT].taxable == events[0].balance.amount + events[1].balance.amount
```

Then it is easier to adjust later and to understand it. Same below

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 2]
```diff
@@ -86,11 +158,21 @@ def test_paraswap_swap_with_fee(accountant: 'Accountant'):
             free_amount=ZERO,
             taxable_amount=FVal(swap_amount_str),
             price=Price(ONE),
-            pnl=PNL(taxable=FVal(swap_amount_str), free=ZERO),
-            cost_basis=CostBasisInfo(taxable_amount=FVal(swap_amount_str), taxable_bought_cost=ZERO, taxfree_bought_cost=ZERO, matched_acquisitions=[], is_complete=False),  # noqa: E501
-            index=0,
+            pnl=PNL(taxable=FVal(swap_amount_str) / 2, free=ZERO),
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:161]: this `/ 2` comes from the price right? Maybe create a variable that points `MOCKED_PRICES[event.asset][event.timestamp]`

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 3]
```diff
@@ -86,11 +158,21 @@ def test_paraswap_swap_with_fee(accountant: 'Accountant'):
...
+            cost_basis=CostBasisInfo(
+                taxable_amount=FVal(swap_amount_str),
+                taxable_bought_cost=FVal(swap_amount_str) / 2,
+                taxfree_bought_cost=ZERO,
+                is_complete=True,
+                matched_acquisitions=[MatchedAcquisition(
+                    taxable=True,
+                    amount=FVal(swap_amount_str),
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:169]:
```suggestion
                    amount=acquisition_event.amount,
```

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 4]
```diff
@@ -73,11 +109,47 @@ def test_paraswap_swap_with_fee(accountant: 'Accountant'):
...
     extra_data = {
-        'group_id': '1' + tx_hash.hex() + '12',  # pylint: disable=no-member
+        'group_id': '1' + tx_hash.hex() + '23',  # pylint: disable=no-member
         'tx_hash': tx_hash.hex(),  # pylint: disable=no-member
     }
+    aquisition_event = AssetAcquisitionEvent(
+        index=0,
+        timestamp=TIMESTAMP_0_SEC,
+        amount=FVal(swap_amount_str),
+        rate=Price(FVal(0.5)),
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:119]: how does the rate play later? Should it be the price?

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 5]
```diff
@@ -73,11 +109,47 @@ def test_paraswap_swap_with_fee(accountant: 'Accountant'):
...
+        ),
+        ProcessedAccountingEvent(
+            index=1,
+            event_type=AccountingEventType.TRANSACTION_EVENT,
+            notes=f'Receive {ONE} USDC from {aquired_from_address} to {user_address}',
+            location=Location.ETHEREUM,
+            timestamp=TIMESTAMP_0_SEC,
+            asset=A_USDC,
+            free_amount=ZERO,
+            taxable_amount=FVal(ONE),
```
Comment (yabirgb) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:145]: here and in the other places ONE is already an Fval. You don't need to FVal it again

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 6]
```diff
@@ -26,10 +36,36 @@
 @pytest.mark.parametrize('accounting_initialize_parameters', [True])
 def test_paraswap_swap_with_fee(accountant: 'Accountant'):
     """Test that the fee in paraswap is handled correctly during accounting"""
-    tx_hash, user_address, contract_address, swap_amount_str, fee_amount_str, receive_amount_str = make_evm_tx_hash(), make_evm_address(), make_evm_address(), '138.2851', '11.308417', '1190.359719'  # noqa: E501
+    tx_hash, user_address, contract_address, aquired_from_address, swap_amount_str, fee_amount_str, receive_amount_str = make_evm_tx_hash(), make_evm_address(), make_evm_address(), make_evm_address(), '138.2851', '11.308417', '1190.359719'  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:39]: Also the numbers are hard. For accounting tests numbers to use should be simple and easy to calculate by hand.

```suggestion
    tx_hash, user_address, contract_address, acquired_from_address, swap_amount_str, fee_amount_str, receive_amount_str = make_evm_tx_hash(), make_evm_address(), make_evm_address(), make_evm_address(), '138.2851', '11.308417', '1190.359719'  # noqa: E501
```

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 7]
```diff
@@ -26,10 +36,36 @@
...
+        sequence_index=0,
+        timestamp=TIMESTAMP_0_MS,  # half price at t=0
+        location=Location.ETHEREUM,
+        event_type=HistoryEventType.RECEIVE,
+        event_subtype=HistoryEventSubType.NONE,
+        asset=A_WBTC,
+        balance=Balance(amount=FVal(swap_amount_str)),
+        location_label=user_address,
+        notes=f'Receive {swap_amount_str} WBTC from {aquired_from_address} to {user_address}',
+        counterparty=CPT_PARASWAP,
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:51]: Hey can you make this into another random address?

Makes the reader think it's actually paraswap related while it's not. Also add a comment this event is to create cost basis for WBTC

```suggestion
        counterparty=0xrandomaddress
```

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 8]
```diff
@@ -26,10 +36,36 @@
...
         sequence_index=1,
+        timestamp=TIMESTAMP_0_MS,  # same price at t=0
+        location=Location.ETHEREUM,
+        event_type=HistoryEventType.RECEIVE,
+        event_subtype=HistoryEventSubType.NONE,
+        asset=A_USDC,
+        balance=Balance(amount=ONE),
+        location_label=user_address,
+        notes=f'Receive {ONE} USDC from {aquired_from_address} to {user_address}',
+        counterparty=CPT_PARASWAP,
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:64]: same here

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 9]
```diff
@@ -26,10 +36,36 @@
...
+        asset=A_WBTC,
+        balance=Balance(amount=FVal(swap_amount_str)),
+        location_label=user_address,
+        notes=f'Receive {swap_amount_str} WBTC from {aquired_from_address} to {user_address}',
+        counterparty=CPT_PARASWAP,
+        address=aquired_from_address,
+    ), EvmEvent(
         tx_hash=tx_hash,
         sequence_index=1,
+        timestamp=TIMESTAMP_0_MS,  # same price at t=0
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:56]: Why do you need this acquisition of 1 USDC?

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 10]
```diff
@@ -73,24 +109,71 @@ def test_paraswap_swap_with_fee(accountant: 'Accountant'):
         pot.events_accountant.process(event=event, events_iterator=events_iterator)  # type: ignore
 
     extra_data = {
-        'group_id': '1' + tx_hash.hex() + '12',  # pylint: disable=no-member
+        'group_id': '1' + tx_hash.hex() + '23',  # pylint: disable=no-member
         'tx_hash': tx_hash.hex(),  # pylint: disable=no-member
     }
+    wbtc_prices = MOCKED_PRICES[A_WBTC.identifier]['EUR']
+    aquisition_event = AssetAcquisitionEvent(
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:116]:
```suggestion
    acquisition_event = AssetAcquisitionEvent(
```

File: rotkehlchen/tests/unit/accounting/evm/test_paraswap.py [hunk 11]
```diff
@@ -114,8 +197,11 @@ def test_paraswap_swap_with_fee(accountant: 'Accountant'):
...
             pnl=PNL(taxable=ZERO, free=ZERO),
             cost_basis=None,
-            index=2,
             extra_data=extra_data,
         )]
     expected_processed_events[0].count_cost_basis_pnl = True  # since it's not settable at ctor
+    expected_processed_events[1].count_cost_basis_pnl = True  # since it's not settable at ctor
+    expected_processed_events[2].count_cost_basis_pnl = True  # since it's not settable at ctor
     assert pot.processed_events == expected_processed_events
+    assert pot.pnls[AccountingEventType.TRANSACTION_EVENT].taxable == events[0].balance.amount + events[1].balance.amount  # noqa: E501
```
Comment (LefterisJP) [inline rotkehlchen/tests/unit/accounting/evm/test_paraswap.py:206]: Why do you think this is correct? So the PnL of this entire action is amount of the first 2 receivals? What about the fee? I do not understand what this test is trying to test and as a reader it's impossible to understand what you are trying to test for

File: rotkehlchen/accounting/history_base_entries.py [hunk 12]
```diff
@@ -196,33 +214,31 @@ def _process_swap(
...
-                given_price=fee_price,
-                # By setting the taxable amount ratio we determine how much of the fee
-                # spending should be a taxable spend and how much free.
-                taxable_amount_ratio=fee_taxable_amount_ratio,
-                count_cost_basis_pnl=True,
-                count_entire_amount_spend=True,
-                extra_data=extra_data,
-            )
-            consumed_events = 3
+            def add_fee_event() -> None:
```
Comment (LefterisJP) [inline rotkehlchen/accounting/history_base_entries.py:217]: same here. And this is exactly why you have to add so many type ignores here ... this is a hack. Undo

File: rotkehlchen/accounting/history_base_entries.py [hunk 13]
```diff
@@ -178,6 +179,20 @@ def _process_swap(
...
+            'notes': in_event.notes if in_event.notes else '',
+            'location': in_event.location,
+            'timestamp': timestamp,
+            'asset': in_event.asset,
+            'amount': in_event.balance.amount,
+            'taxable': False,  # acquisitions in swaps are never taxable
+            'given_price': prices[1],
+            'extra_data': extra_data,
+        }
+
```
Comment (LefterisJP) [inline rotkehlchen/accounting/history_base_entries.py:194]:
```suggestion
```

File: rotkehlchen/accounting/history_base_entries.py [hunk 14]
```diff
@@ -196,33 +211,30 @@ def _process_swap(
...
-            asset=in_event.asset,
-            amount=in_event.balance.amount,
-            taxable=False,  # acquisitions in swaps are never taxable
-            given_price=prices[1],
-            extra_data=extra_data,
-        )
-        return consumed_events
+            events_to_add_queue.extend([
+                ('add_in_event', add_in_event_kwargs),
+                ('add_out_event', {
```
Comment (LefterisJP) [inline rotkehlchen/accounting/history_base_entries.py:216]: no need for dynamic names. It's constant

```suggestion
                (self.pot.add_in_event, add_in_event_kwargs),
                (self.pot.add_out_event, {
```

File: rotkehlchen/accounting/history_base_entries.py [hunk 15]
```diff
@@ -196,33 +211,30 @@ def _process_swap(
...
+                    'extra_data': extra_data,
+                }),
+            ])
+            if fee_event.asset == out_event.asset or fee_event.sequence_index < in_event.sequence_index:  # noqa: E501
+                events_to_add_queue.reverse()  # we add fee first
+        else:
+            events_to_add_queue.append(('add_in_event', add_in_event_kwargs))
+
+        for adder_function, kwargs in events_to_add_queue:
+            getattr(self.pot, adder_function)(**kwargs)  # add the queued events
```
Comment (LefterisJP) [inline rotkehlchen/accounting/history_base_entries.py:239]:
```suggestion
            events_to_add_queue.append((self.pot.add_in_event, add_in_event_kwargs))

        for adding_method, kwargs in events_to_add_queue:
            adding_method(**kwargs)  # add the queued events
```

File: rotkehlchen/accounting/history_base_entries.py [hunk 16]
```diff
@@ -196,33 +211,30 @@ def _process_swap(
...
+        return 1 + len(events_to_add_queue)  # we consumed in/out
```
Comment (LefterisJP) [inline rotkehlchen/accounting/history_base_entries.py:240]:
```suggestion
        return 1 + len(events_to_add_queue)  # consumed events
```

**Review by LefterisJP — APPROVED** [review]
LGTM. Failed tests passed locally and seem to hit timeouts. Perhaps we should adjust them or VCR them.

**Pull request merged**